quickwit-config/src/config.rs

```diff
@@ -280,69 +280,77 @@ impl QuickwitConfig {
+            Ok(ip_addr) => get_socket_addr(&(ip_addr, self.grpc_listen_port())).await,
+            Err(_) => self.grpc_listen_addr().await,
         }
     }
 
-    /// Returns the grpc socket address.
-    pub fn grpc_socket_addr(&self) -> anyhow::Result<SocketAddr> {
-        get_socket_addr(&(self.listen_address.as_str(), self.grpc_listen_port()))
+    /// Returns the gossip listen port of the node (UDP).
+    /// Extracted from the config if specified or same as `rest_listen_port` otherwise.
+    fn gossip_listen_port(&self) -> u16 {
+        // By default, we use the same port number as the REST port but UDP this time.
+        self.gossip_listen_port.unwrap_or(self.rest_listen_port)
     }
 
-    /// Returns the gossip socket address.
-    pub fn gossip_socket_addr(&self) -> anyhow::Result<SocketAddr> {
-        get_socket_addr(&(self.listen_address.as_str(), self.gossip_listen_port()))
+    /// Returns the gossip listen address of the node, i.e. the UDP socket address on which the node
+    /// receives gossip messages.
+    pub async fn gossip_listen_addr(&self) -> anyhow::Result<SocketAddr> {
+        get_socket_addr(&(self.listen_address.as_str(), self.gossip_listen_port())).await
     }
 
-    /// The node gossip_public_address should ideally be specified via config;
-    /// environment variable interpolation if necessary.
-    /// right now we just try to fetch from environment variable otherwise fallback
-    /// to listen_address.
-    pub fn gossip_public_addr(&self) -> anyhow::Result<SocketAddr> {
-        match env::var("QW_GOSSIP_PUBLIC_ADDRESS") {
-            Ok(addr) => Ok(addr.parse::<SocketAddr>()?),
-            Err(_) => self.gossip_socket_addr(),
+    /// Returns the gossip public address of the node, i.e. the socket address to send UDP packets
+    /// to in order to gossip with the node.
+    pub async fn gossip_public_addr(&self) -> anyhow::Result<SocketAddr> {
+        match env::var("QW_PUBLIC_IP") {
+            Ok(ip_addr) => get_socket_addr(&(ip_addr, self.gossip_listen_port())).await,
+            Err(_) => self.gossip_listen_addr().await,
         }
     }
 
-    pub fn seed_socket_addrs(&self) -> anyhow::Result<Vec<SocketAddr>> {
-        // If no port is given, we assume a seed is using the same port as ourself.
-        let default_gossip_port = self.gossip_socket_addr()?.port();
-        let seed_socket_addrs: Vec<SocketAddr> = self
-            .peer_seeds
-            .iter()
-            .flat_map(|seed| {
-                parse_socket_addr_with_default_port(seed, default_gossip_port)
-                    .map_err(|_| warn!(address = %seed, "Failed to resolve seed address."))
-            })
-            .collect();
+    /// Returns the list of peer seed addresses. The addresses MUST NOT be resolved. Otherwise, the
+    /// DNS-based discovery mechanism implemented in Chitchat will not work correctly.
+    pub async fn peer_seed_addrs(&self) -> anyhow::Result<Vec<String>> {
```

Comment:
We want to pass non-resolved addresses to Chitchat but still want to resolve them for validation. Additionally, we need to append a default port if necessary and finally return the addresses as strings, which is tricky for Ipv6. That's why I introduced the `HostAddr` struct to deal with the whole thing.

Reply:
Thks! This is a great explanation. I think you should put it as comment in the code I think.

Reply:
Comment added.